async-nats/src/header.rs

```diff
@@ -12,6 +12,199 @@
 // limitations under the License.
 
 //! NATS [Message][crate::Message] headers, leveraging [http::header] crate.
-pub use http::header::{HeaderMap, HeaderName, HeaderValue};
+// pub use http::header::{HeaderMap, HeaderName, HeaderValue};
+
+use std::{
+    collections::{self, HashMap, HashSet},
+    str::FromStr,
+};
+
+use serde::Serialize;
 
 pub const NATS_LAST_STREAM: &str = "nats-last-stream";
+
+#[derive(Clone, PartialEq, Eq, Debug, Serialize, Default)]
+pub struct HeaderMap {
+    headers: HashMap<HeaderName, HeaderValue>,
+}
+
+impl FromIterator<(HeaderName, HeaderValue)> for HeaderMap {
+    fn from_iter<T: IntoIterator<Item = (HeaderName, HeaderValue)>>(iter: T) -> Self {
+        let mut header_map = HeaderMap::new();
+        for (key, value) in iter {
+            header_map.insert(key, value);
+        }
+        header_map
+    }
+}
+
+impl HeaderMap {
+    pub fn iter(&self) -> std::collections::hash_map::Iter<'_, HeaderName, HeaderValue> {
+        self.headers.iter()
+    }
+}
+
+impl HeaderMap {
+    pub fn new() -> Self {
+        HeaderMap::default()
+    }
+}
+
+impl HeaderMap {
+    pub fn insert<K: IntoHeaderName, V: IntoHeaderValue>(&mut self, name: K, value: V) {
+        self.headers
+            .insert(name.into_header_name(), value.into_header_value());
+    }
+
+    pub fn append<K: IntoHeaderName, V: ToString>(&mut self, name: K, value: V) {
+        let key = name.into_header_name();
+        let v = self.headers.get_mut(&key);
+        match v {
+            Some(v) => {
+                v.value.insert(value.to_string());
+            }
+            None => {
+                self.insert(key, value.to_string().into_header_value());
+            }
+        }
+    }
+
+    pub fn get<T: IntoHeaderName>(&self, name: T) -> Option<&HeaderValue> {
+        self.headers.get(&name.into_header_name())
+    }
+
+    pub fn to_bytes(&self) -> Vec<u8> {
```

Comment:
Should serialization/deserialization (from_str) be here, or with the rest of protocol stuff?

Reply:
That's a good question I tried to answer myself.

It feels like it belongs to protocol stuff, but at the same time, header serialisation is quite specific to headers only, especially NATS headers.

Reply:
Alright I'll revisit this next week or so. Lets merge and get KV out.